File: aligny.py
import numpy as np
import matplotlib.pyplot as plt
#import scipy
from scipy.optimize import minimize
#from scipy.optimize import NonlinearConstraint
from anchormean0to1 import anchormean0to1

def aligny(G_0):
    
    print("Beginning Y alignment")
    
    N = G_0.shape[1]
    alpha_0 = np.zeros([1,N])
    beta_0 = np.ones([1,N])
    G_0 = (G_0 - alpha_0)/beta_0
    G_mean_0 = np.nanmean(G_0,1)
    

    p_0 = np.squeeze(np.array([np.transpose(alpha_0),np.transpose(beta_0)]))
    #p_0 = ([[alpha_0],[beta_0]])
    
    # alpha lower bound is -1, upper bound is 1
    # beta lower bound is 0,   upper bound is inf

    lb_alpha = np.ones([N,1])*-1
    lb_beta = np.zeros([N,1])
    ub_alpha = np.ones([N,1])
    ub_beta = np.ones([N,1])*np.inf
    
    lb = np.vstack([lb_alpha, lb_beta])
    ub = np.vstack([ub_alpha, ub_beta])
    
    bnds = np.hstack([lb,ub])
    
    ### Working to here -MJT
    fun = lambda params: chisq(params, G_0, G_mean_0, N)
    solution = minimize(fun, p_0, method='SLSQP', bounds=bnds, constraints=(), 
                        tol=1e-6, callback=None, options=None)
    #                    options={'gtol': 1e-6, 'disp': True})
    
    #solution = minimize(chisq(p_0,G_0,G_mean_0), p_0, method='SLSQP', bounds=None, constraints=(), tol=None, callback=None, options=None)
    
    # -> line 618 _minimize.py -> line 399 slsqp.py -> line 327 optimize.py -> line 26 aligny.py -> line 52 aligny.py
    # *** IndexError: too many indices for array
    
    #f = chisq(p_0, G_0, G_mean_0)
    #[p,chi2,exitFlag,~] = fmincon(f,p_0,[],[],[],[],lb,ub,[],options);
    
    #scipy.optimize.NonLinearConstraint(f,lb,ub)


    alpha = solution.x[0:N]
    beta = solution.x[N:2*N]

    g = (G_0 - alpha)/beta
    
    return g



#Cost Function

def chisq(params,G,g_mean,N):
    alpha = params[0:N]
    beta = params[N:2*N]
    chi2 = np.nanmean(np.nansum ((G - (alpha + beta))**2))
    
    return chi2

File: aligny.m
function [y, alpha, beta, chi2] = aligny(Y_0)
%ALIGNY aligns data along the Y-axis of data in a 2D matrix.
% 
% Input:
%   > Y_0: initial profile data. 2D (no. data points) x (no. replicates).
%
% Output:
%   > y: aligned output data. 2D. Same size as input.
%   > alpha: additive parameter (1 per profile)
%   > beta: scaling/multiplicative parameter (1 per profile)
%   > chi2: final squared error

disp('Beginning Y alignment')

[nPoints, nReplicates] = size(Y_0);

%% Initialize alignment parameters.
alpha_0 = zeros(1,nReplicates);
beta_0 = ones(1,nReplicates);

Y_0 = (Y_0 - alpha_0)./beta_0;
Y_mean_0 = nanmean(Y_0,2);

p_0 = [alpha_0;beta_0];

%% Set optimization options.
options = optimset( ...
    'MaxIter', 1e6, ...
    'MaxFunEvals', 1e6, ...
    'Display','notify-detailed');

%% Set optimization parameter constraints.
lb = [-ones(1,nReplicates);zeros(1,nReplicates)];
ub = [ones(1,nReplicates);inf(1,nReplicates)];

%% Optimization.
f = @(p_0)chisq(p_0, Y_0, Y_mean_0);
[p,chi2,exitFlag,~] = fmincon(f,p_0,[],[],[],[],lb,ub,[],options);

exitFlag

%% Apply final optimization parameters to profile data.
alpha = p(1,:);
beta = p(2,:);

y = (Y_0 - alpha)./beta;

end

%% Cost function
function chi2 = chisq(p,Y,y_mean)
   alpha = p(1,:);
   beta = p(2,:);
   chi2 = nanmean(nansum( (Y - (alpha + beta.*y_mean)).^2 ));
end